drivers/platform/mbed/mbed_uart.cpp

```diff
@@ -0,0 +1,492 @@
+		usb_cdc_dev = (platform_usbcdc *)((struct mbed_uart_desc *)(
+				desc->extra))->uart_port;
+
+		while (bytes_number) {
+			while (!usb_cdc_dev->data_transmited()) {
+				/* Wait until old data is transmitted */
+			}
+
+			/* Make sure packet size is less than max CDC packet size during data transmit */
+			d_sz = (bytes_number > (USB_CDC_MAX_PACKET_SIZE - 1)) ?
+			       (USB_CDC_MAX_PACKET_SIZE - 1) :
+			       bytes_number;
+
+			/* Change terminal connection status manually */
+			usb_cdc_dev->change_terminal_connection(true);
+
+			usb_cdc_dev->send_nb((uint8_t *)&data[indx], d_sz, &d_sz);
+
+			bytes_number -= d_sz;
+			indx += d_sz;
+		}
+
+		return tx_bytes;
+	} else {
+		uart = (BufferedSerial *)(((struct mbed_uart_desc *)(
+						   desc->extra))->uart_port);
+		return uart->write(data, bytes_number);
+	}
+}
+
+/**
+ * @brief Submit reading buffer to the UART driver.
+ * Buffer is used until bytes_number bytes are read.
+ * @param desc[in]- Descriptor of the UART device
+ * @param data[out] - Buffer where data will be read
+ * @param bytes_number[in] - Number of bytes to be read.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ * @note Currently implemented only for UART and not for USBCDC (VCOM)
+ */
+int32_t uart_read_nonblocking(struct uart_desc *desc,
+			      uint8_t *data,
+			      uint32_t bytes_number)
+{
+	mbed::BufferedSerial *uart;	// pointer to BufferedSerial/UART instance
+	uint32_t i;
+
+	if (!desc || !desc->extra || !data)
+		return -EINVAL;
+
+	if (!((struct mbed_uart_desc *)(desc->extra))->uart_port)
+		return -EINVAL;
+
+	uart = (BufferedSerial *)(((struct mbed_uart_desc *)(
+					   desc->extra))->uart_port);
+
+	/* Return 0 bytes if nothing to read */
+	if (!uart->readable())
+		return 0;
+
+	for (i = 0; i < bytes_number; i++) {
+		if (uart->readable())
+			uart->read(&data[i], 1);
+	}
+
+	return bytes_number;
+}
+
+/**
+ * @brief Submit writting buffer to the UART driver.
+ * Data from the buffer is sent over the UART, the function returns immediatly.
+ * @param desc[in] - Descriptor of the UART device
+ * @param data[in,out] - Buffer where data will be written
+ * @param bytes_number[in] - Number of bytes to be written.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ * @note Currently implemented only for UART and not for USBCDC (VCOM)
+ */
+int32_t uart_write_nonblocking(struct uart_desc *desc,
+			       const uint8_t *data,
+			       uint32_t bytes_number)
+{
+	mbed::BufferedSerial *uart;		// pointer to BufferedSerial/UART instance
+	uint32_t i;
+
+	if (!desc || !desc->extra || !data)
+		return - EINVAL;
+
+	if (!((struct mbed_uart_desc *)(desc->extra))->uart_port)
+		return -EINVAL;
+
+	uart = (BufferedSerial *)(((struct mbed_uart_desc *)(
+					   desc->extra))->uart_port);
+
+	/* Return error if unable to write */
+	if (!uart->writable())
+		return -EIO;
+
+	for (i = 0; i < bytes_number; i++) {
+		if (uart ->writable())
+			uart->write(&data[i], 1);
```

Comment:
I'd say things are easier here. We just have to account for the maximum number of bytes that can be queued to be written at once. If `writable`, compare `bytes_number` with the maximum that can be queued at once and queue the smallest number. Then return the number of bytes queued. The for shouldn't be necessary here.

Reply:
Not sure if I get you correctly here. 
Do you mean:
if (uart->writable())
		return uart->write(data, bytes_number);
else
		return 0;

Reply:
Yes, this ca work, but does `uart->write` have a maximum number of bytes that can be transmitted at once? From the documentation it seems that it will automatically write a partial write if the requested transfer is too big. So we can do just:

if (uart->writable())
return uart->write(data, bytes_number);

Or it might even work without `if` because if it can't write anything it will return `-EAGAIN`. I'd say both can work.
Note: `uart->write` will return the number of bytes it transmitted, or queued, even if it's just a partial transfer.

Reply:
Okay. I will implement the first one. I was also thinking, we can do same for non-blocking read. If ```readable()``` is true, we will just return ```read()``` with number of bytes else will return 0.
e.g. 
```
if (readable())
    return read(data, bytes)
else
   return 0
```
Let me know if this is okay to implement

Reply:
This seems good. Thank you!